# Keras Basics
In this notebook we will learn basics of Keras with the help a data set known as iris. In this dataset we have few features of flowers on basis of these features we have to predict three classes( 3 flower species )

## Import the tools and Load data

In [93]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris

In [94]:
iris=load_iris()

In [101]:
print(type(iris))

<class 'sklearn.utils.Bunch'>


In [103]:
print(iris.DESCR)

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
                
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
    ============== ==== ==== ======= ===== ===========

In [104]:
X=iris.data

In [107]:
print(iris.feature_names)

['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


In [108]:
y=iris.target

In [109]:
len(X)

150

In [110]:
len(y)

150

## Change y using One hot encoding

In [111]:
from keras.utils import to_categorical
y=to_categorical(y)

In [112]:
y.shape

(150, 3)

In [114]:
y[0:5]

array([[1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.]], dtype=float32)

## Split the data into training and test datasets

In [115]:
from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.25,random_state=42)

In [116]:
X_train.shape

(112, 4)

In [117]:
y_train.shape

(112, 3)

In [118]:
X_train[:3]

array([[5. , 3.6, 1.4, 0.2],
       [5.2, 4.1, 1.5, 0.1],
       [5.8, 2.7, 5.1, 1.9]])

## Standardize the datasets

In [119]:
from sklearn.preprocessing import MinMaxScaler
scaler_object=MinMaxScaler()
scaler_object.fit(X_train)

scaled_X_train=scaler_object.transform(X_train)
scaled_X_test=scaler_object.transform(X_test)

In [120]:
scaled_X_test[:10]

array([[0.52941176, 0.36363636, 0.64285714, 0.45833333],
       [0.41176471, 0.81818182, 0.10714286, 0.08333333],
       [1.        , 0.27272727, 1.03571429, 0.91666667],
       [0.5       , 0.40909091, 0.60714286, 0.58333333],
       [0.73529412, 0.36363636, 0.66071429, 0.54166667],
       [0.32352941, 0.63636364, 0.07142857, 0.125     ],
       [0.38235294, 0.40909091, 0.44642857, 0.5       ],
       [0.76470588, 0.5       , 0.71428571, 0.91666667],
       [0.55882353, 0.09090909, 0.60714286, 0.58333333],
       [0.44117647, 0.31818182, 0.5       , 0.45833333]])

## Let's build a keras model from the scratch

In [122]:
from keras.models import Sequential
from keras.layers import Dense,Dropout

In [134]:
model=Sequential()
# Input layer
model.add(Dense(8,input_dim=4,activation="relu"))
model.add(Dense(32,input_dim=4,activation='relu'))
model.add(Dropout(0.35))
model.add(Dense(3,activation='softmax'))
model.compile(loss="categorical_crossentropy",optimizer="adam",metrics=['accuracy'])

model.summary()

Model: "sequential_6"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_16 (Dense)             (None, 8)                 40        
_________________________________________________________________
dense_17 (Dense)             (None, 32)                288       
_________________________________________________________________
dropout_5 (Dropout)          (None, 32)                0         
_________________________________________________________________
dense_18 (Dense)             (None, 3)                 99        
Total params: 427
Trainable params: 427
Non-trainable params: 0
_________________________________________________________________


In [135]:
model.fit(scaled_X_train,y_train,epochs=200,verbose=2)

Epoch 1/200
 - 0s - loss: 1.0252 - accuracy: 0.3482
Epoch 2/200
 - 0s - loss: 0.9946 - accuracy: 0.4286
Epoch 3/200
 - 0s - loss: 0.9820 - accuracy: 0.3750
Epoch 4/200
 - 0s - loss: 1.0043 - accuracy: 0.3304
Epoch 5/200
 - 0s - loss: 0.9882 - accuracy: 0.3214
Epoch 6/200
 - 0s - loss: 0.9682 - accuracy: 0.3929
Epoch 7/200
 - 0s - loss: 0.9831 - accuracy: 0.3304
Epoch 8/200
 - 0s - loss: 0.9655 - accuracy: 0.4196
Epoch 9/200
 - 0s - loss: 0.9507 - accuracy: 0.4375
Epoch 10/200
 - 0s - loss: 0.9367 - accuracy: 0.4107
Epoch 11/200
 - 0s - loss: 0.9206 - accuracy: 0.4643
Epoch 12/200
 - 0s - loss: 0.9329 - accuracy: 0.4643
Epoch 13/200
 - 0s - loss: 0.9132 - accuracy: 0.4643
Epoch 14/200
 - 0s - loss: 0.9192 - accuracy: 0.5000
Epoch 15/200
 - 0s - loss: 0.8845 - accuracy: 0.6071
Epoch 16/200
 - 0s - loss: 0.8794 - accuracy: 0.6071
Epoch 17/200
 - 0s - loss: 0.8769 - accuracy: 0.5446
Epoch 18/200
 - 0s - loss: 0.8695 - accuracy: 0.6071
Epoch 19/200
 - 0s - loss: 0.8560 - accuracy: 0.6429
Ep

Epoch 155/200
 - 0s - loss: 0.2118 - accuracy: 0.9196
Epoch 156/200
 - 0s - loss: 0.1956 - accuracy: 0.9643
Epoch 157/200
 - 0s - loss: 0.2089 - accuracy: 0.9554
Epoch 158/200
 - 0s - loss: 0.2051 - accuracy: 0.9554
Epoch 159/200
 - 0s - loss: 0.1884 - accuracy: 0.9643
Epoch 160/200
 - 0s - loss: 0.2102 - accuracy: 0.9464
Epoch 161/200
 - 0s - loss: 0.2213 - accuracy: 0.9196
Epoch 162/200
 - 0s - loss: 0.2064 - accuracy: 0.9464
Epoch 163/200
 - 0s - loss: 0.2008 - accuracy: 0.9732
Epoch 164/200
 - 0s - loss: 0.2142 - accuracy: 0.9643
Epoch 165/200
 - 0s - loss: 0.2068 - accuracy: 0.9643
Epoch 166/200
 - 0s - loss: 0.2218 - accuracy: 0.9196
Epoch 167/200
 - 0s - loss: 0.2031 - accuracy: 0.9375
Epoch 168/200
 - 0s - loss: 0.2039 - accuracy: 0.9286
Epoch 169/200
 - 0s - loss: 0.2189 - accuracy: 0.9464
Epoch 170/200
 - 0s - loss: 0.2085 - accuracy: 0.9375
Epoch 171/200
 - 0s - loss: 0.1708 - accuracy: 0.9643
Epoch 172/200
 - 0s - loss: 0.1806 - accuracy: 0.9554
Epoch 173/200
 - 0s - loss: 

## Evaluate the model

In [137]:
model.evaluate(scaled_X_test,y_test)

38/38 [==============================] - 0s 2ms/step


[0.11282200052549965, 0.9736841917037964]

In [140]:
model.metrics_names

['loss', 'accuracy']

In [141]:
predictions=model.predict_classes(scaled_X_test)

In [142]:
from sklearn.metrics import classification_report,confusion_matrix

In [143]:
print(classification_report(y_test.argmax(axis=1),predictions))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        15
           1       1.00      0.91      0.95        11
           2       0.92      1.00      0.96        12

   micro avg       0.97      0.97      0.97        38
   macro avg       0.97      0.97      0.97        38
weighted avg       0.98      0.97      0.97        38



In [144]:
confusion_matrix(y_test.argmax(axis=1),predictions)

array([[15,  0,  0],
       [ 0, 10,  1],
       [ 0,  0, 12]], dtype=int64)

In [145]:
df=pd.DataFrame(confusion_matrix(y_test.argmax(axis=1),predictions),index=[0,1,2],columns=[0,1,2])
df

0   1   2
0  15   0   0
1   0  10   1
2   0   0  12

## So our Keras model is working fantastically with an accuracy of around 97% which is great to start with